File: book-service/src/main/java/com/bookstore/books/book/entity/Book.java
package com.bookstore.books.book.entity;
import com.bookstore.books.author.entity.Author;
import com.bookstore.books.category.entity.Category;
import com.bookstore.books.publisher.entity.Publisher;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "books")
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(length = 36)
    private UUID id;

    @Column(unique = true, nullable = false)
    private String isbn;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    private BigDecimal price;

    private String language;

    private LocalDate publishedDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "publisher_id")
    private Publisher publisher;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "book_authors",
            joinColumns = @JoinColumn(name = "book_id"),
            inverseJoinColumns = @JoinColumn(name = "author_id")
    )
    @Builder.Default
    private Set<Author> authors = new HashSet<>();

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: frontend/src/types/book.ts
export interface Book {
  id: string;
  isbn: string;
  title: string;
  description: string | null;
  price: number;
  language: string;
  publishedDate: string | null;
  coverImageUrl: string | null;
  publisherId: string;
  categoryId: string;
  authorIds: string[];
}

export interface BookCreateRequest {
  isbn: string;
  title: string;
  description?: string;
  price: number;
  language: string;
  publishedDate?: string;
  publisherId: string;
  categoryId: string;
  authorIds: string[];
}

export interface Author {
  id: string;
  firstName: string;
  lastName: string;
  biography: string | null;
  country: string | null;
}

export interface Category {
  id: string;
  name: string;
  description: string | null;
}

export interface Publisher {
  id: string;
  name: string;
  address: string | null;
}
